py/server/deephaven/liveness_scope.py

```diff
@@ -27,44 +50,136 @@ def __init__(self, j_scope: jpy.JType):
-                self.j_scope.release()
-                self.j_scope = None
-            except Exception as e:
-                raise DHError(e, message="failed to close the LivenessScope.")
+        warn('This function is deprecated, prefer release(). Use cases that rely on this are likely to now fail.',
+             DeprecationWarning, stacklevel=2)
+        _pop(self.j_scope)
+        self.release()
+
+    @contextlib.contextmanager
+    def open(self, release_after_block: bool = True) -> None:
+        """
+        Uses this scope for the duration of the `with` block. The scope defaults to being closed
+        when the block ends, disable by passing release_after_block=False
+
+        Args:
+            release_after_block: True to release the scope when the block ends, False to leave the
+            scope open for reuse.
+
+        Returns: None, to allow changes in the future.
+
+        """
+        _push(self.j_scope)
+        try:
+            yield None
+        finally:
+            _pop(self.j_scope)
+            if release_after_block:
+                self.release()
 
     @property
     def j_object(self) -> jpy.JType:
         return self.j_scope
 
-    def preserve(self, wrapper: JObjectWrapper) -> None:
+    def preserve(self, referent: Union[JObjectWrapper, jpy.JType]) -> None:
         """Preserves a query graph node (usually a Table) to keep it live for the outer scope.
 
         Args:
-            wrapper (JObjectWrapper): a wrapped Java object such as a Table
+            referent (Union[JObjectWrapper, jpy.JType]): an object to preserve in the next outer liveness scope
 
         Raises:
             DHError
         """
+        referent = _unwrap_to_liveness_referent(referent)
         try:
             _JLivenessScopeStack.pop(self.j_scope)
-            _JLivenessScopeStack.peek().manage(wrapper.j_object)
-            _JLivenessScopeStack.push(self.j_scope)
+            _JLivenessScopeStack.peek().manage(_unwrap_to_liveness_referent(referent))
         except Exception as e:
             raise DHError(e, message="failed to preserve a wrapped object in this LivenessScope.")
+        finally:
+            _JLivenessScopeStack.push(self.j_scope)
+
+    def manage(self, referent: Union[JObjectWrapper, jpy.JType]) -> None:
+        """
+        Explicitly manage the given java object in this scope. Must only be passed a Java LivenessReferent, or
+        a Python wrapper around a LivenessReferent
+
+        Args:
+            referent: the object to manage by this scope
+
+        Returns: None
+
+        Raises:
+            DHError if the referent isn't a LivenessReferent, or if it is no longer live
+
+        """
+        referent = _unwrap_to_liveness_referent(referent)
+        try:
+            self.j_scope.manage(referent)
+        except Exception as e:
+            raise DHError(e, message="failed to manage object")
+
+    def unmanage(self, referent: Union[JObjectWrapper, jpy.JType]) -> None:
```

Comment:
Maybe we should be wrapping `tryManage` and `tryUnmanage`, too. Arguably, they are more useful inside of Java implementation classes.

Reply:
Yeah I wasn't sure if we should have these anyway - mostly meant as an easier (and doc'd) escape hatch for NextTimeThisHappens, rather than sending the user to figure out jpy fun As jpy goes, this isn't too bad at least.

Reply:
I'm not sure what we're concluding. `tryManage` and `tryUnmanage` exist, but mostly for cases internal to the engine/operations where we have race conditions around whether a user may have caused an object's reference count to reach zero while we're concurrently doing things with the object. Meaning, I'd be OK:
1. leaving this as-is and not wrapping those methods at this time, or
2. wrapping them now to avoid future work.

Reply:
We decided on #1.